lib/Dialect/FIRRTL/Transforms/CreateSiFiveMetadata.cpp

```diff
@@ -69,12 +77,60 @@ LogicalResult CreateSiFiveMetadataPass::emitMemoryMetadata() {
       !dutMod ||
       instancePathCache.instanceGraph.getTopLevelNode()->getModule() == dutMod;
 
+  auto *context = circuitOp.getContext();
+  auto unknownLoc = mlir::UnknownLoc::get(context);
+  auto builderOM =
+      mlir::ImplicitLocOpBuilder::atBlockEnd(unknownLoc, moduleOp.getBody());
+
+  // Add all the properties of a memory as fields of the class.
+  // The types must match exactly with the FMemModuleOp attribute type.
+
+  StringRef paramNames[] = {"name",           "depth",        "width",
+                            "maskBits",       "readPorts",    "writePorts",
+                            "readwritePorts", "writeLatency", "readLatency"};
+  mlir::Type classFieldTypes[] = {
+      om::SymbolRefType::get(context),
+      mlir::IntegerType::get(context, 64, IntegerType::Unsigned),
+      mlir::IntegerType::get(context, 32, IntegerType::Unsigned),
+      mlir::IntegerType::get(context, 32, IntegerType::Unsigned),
+      mlir::IntegerType::get(context, 32, IntegerType::Unsigned),
+      mlir::IntegerType::get(context, 32, IntegerType::Unsigned),
+      mlir::IntegerType::get(context, 32, IntegerType::Unsigned),
+      mlir::IntegerType::get(context, 32, IntegerType::Unsigned),
+      mlir::IntegerType::get(context, 32, IntegerType::Unsigned)};
+
+  om::ClassOp memorySchemaClass;
+  auto createMemoryOMSchema = [&]() {
+    // Memory metadata class.
+    memorySchemaClass =
+        builderOM.create<circt::om::ClassOp>("MemorySchema", paramNames);
+
+    // Now add all the properties of the memory as a ClassFieldOp.
+    Block *body = new Block();
+    memorySchemaClass.getRegion().push_back(body);
+    builderOM.setInsertionPointToEnd(body);
+    for (auto [fieldName, fieldType] : llvm::zip(paramNames, classFieldTypes))
+      builderOM.create<om::ClassFieldOp>(
+          fieldName, body->addArgument(fieldType, unknownLoc));
+
+    // Now create the class that will instantiate metadata class with all the
+    // memories of the circt.
+    builderOM.setInsertionPointToEnd(moduleOp.getBody());
+    auto metadataClass = builderOM.create<circt::om::ClassOp>("MemoryMetadata");
+    auto *memMetadataBlock = new Block();
+    metadataClass.getRegion().push_back(memMetadataBlock);
+    builderOM.setInsertionPointToEnd(memMetadataBlock);
+  };
+  // index will be used to name unique symbols corresponding to each memory.
+  unsigned index = 0;
   // This lambda, writes to the given Json stream all the relevant memory
   // attributes. Also adds the memory attrbutes to the string for creating the
   // memmory conf file.
   auto createMemMetadata = [&](FMemModuleOp mem,
                                llvm::json::OStream &jsonStream,
                                std::string &seqMemConfStr) {
+    if (!memorySchemaClass)
+      createMemoryOMSchema();
```

Comment:
Why do we need lazily compute the OMSchema instead of just initializing memorySchemaClass?

Reply:
relying on lazy creation, so that the om schema classes are only created if at least one memory exists. So, trying to avoid empty om classes.